### ISSUE TITLE
Rack interface with proxy?

### ISSUE THREAD

**Issue by scottjehl** [h2o/h2o#559]
I expected the following to work but it does not appear to set any link headers in the final response:

```
paths:
      /:
        proxy.reverse.url: "http://mywebsite.com/"
        mruby.handler: |
          Proc.new do |env|
            push_paths = []
            if env["PATH_INFO"] == "/"
              push_paths << "/css/dist/all.css"
            end
            [399, push_paths.empty? ? {} : {"link" => push_paths.map{|p| "<#{p}>; rel=preload"}.join("\n")}, []]
          end
```

Does mruby work in addition to proxy or only with file.dir?

**Comment by scottjehl** [comment]
This may have been related to proxying an external domain.

**Comment by kazuho** [comment]
At the moment, the handlers that recognize the `link: rel=preload` header removes such headers from the response, and (if possible) pushes the specified resource to the client.  We may change the behavior in future to not remove them.

> Does mruby work in addition to proxy or only with file.dir?

It should.  One way to confirm is to change the name of the header emitted by mruby from `link` to something else (so that it would not be removed).

**Comment by kazuho** [comment]
@scottjehl 
Revisiting this issue, it seems that the two directives (proxy.handler.url and mruby.handler) are in the wrong order in your configuration file.

When receving an HTTP request, H2O traverses the handlers in the orders they are listed until one handler returns a response.  In case of mruby handler, response is returned unless the returned status code is 399.  In case of proxy handler, a response is _always_ returned (therefore it is always the last handler that is executed).

In short, you should write mruby.handler that returns a 399 status code _above_ the proxy configuration in order to modify the response headers being sent (and/or to trigger HTTP/2 push).

**Comment by kuriyama** [comment]
If order of handlers is important, should this level of configuration file be sequence, not mapping?

I'm generating h2o YAML configuration file from another data structure, so it is impossible to control order of mapping (hash) data...

**Comment by kazuho** [comment]
@kuriyama Thank you for the suggestion.  I think we can support using a sequence in place of mapping _as an option_, without breaking compatibility.

**Comment by kazuho** [comment]
@kuriyama We believe your issue has been resolved in the PR above. Please let us know if you have any problem.
